src/Microsoft.AspNet.OData.Shared/EdmChangedObjectCollectionOfT.cs

```diff
@@ -0,0 +1,297 @@
+                catch
+                {
+                    //For handling the failed operations.
+                    HandleFailedOperation(changedObj, operation, ref edmChangedObject, originalObj);
+                }
+
+                edmChangedObjectCollection.Add(edmChangedObject);
+            }
+
+            return edmChangedObjectCollection;
+        }
+               
+
+        private void HandleFailedOperation(dynamic changedObj, DataModificationOperationKind operation, ref IEdmChangedObject edmChangedObject, TStructuralType originalObj)
+        {
+            DataModificationExceptionType dataModificationExceptionType = new DataModificationExceptionType(operation);
+
+            // This handles the Data Modification exception. This adds Core.DataModificationException annotation and also copy other instance annotations.
+            //The failed operation will be based on the protocol
+            switch (operation)
+            {
+                case DataModificationOperationKind.Update:
+                    edmChangedObject = changedObj;
+                    break;
+                case DataModificationOperationKind.Insert:
+                    {
+                        Type type = typeof(EdmDeltaDeletedEntityObject<>).MakeGenericType(changedObj.ExpectedClrType);
+                        EdmDeltaDeletedEntityObject edmDeletedObject  = Activator.CreateInstance(type, EntityType) as EdmDeltaDeletedEntityObject;
+                        object contentId = changedObj.TransientInstanceAnnotationContainer.GetResourceAnnotation("Core.ContentID");
+                        if(contentId != null)
+                        {
+                            edmDeletedObject.Id = contentId.ToString();                            
+                        }
+                        else
+                        {
+                            edmDeletedObject.Id = string.Empty;
+                        }
+
+                        CopyInstanceAnnotations(changedObj, edmDeletedObject);
+                        edmDeletedObject.TransientInstanceAnnotationContainer.AddResourceAnnotation("Core.DataModificationException", dataModificationExceptionType);                       
+                        edmChangedObject = edmDeletedObject;
+                        break;
+                    }
+                case DataModificationOperationKind.Delete:
+                case DataModificationOperationKind.Unlink:
+                    {
+                        Type type = typeof(EdmDeltaEntityObject<>).MakeGenericType(typeof(TStructuralType));
+                        EdmDeltaEntityObject edmDeltaEntityObject = Activator.CreateInstance(type, EntityType) as EdmDeltaEntityObject;
+                        foreach(PropertyInfo property in typeof(TStructuralType).GetProperties())
+                        {
+                            edmDeltaEntityObject.TrySetPropertyValue(property.Name, property.GetValue(originalObj));
+                        }
```

Comment:
This doesn't handle things like hidden properties or dynamic properties.  Look at the logic in Delta<T> for copying changed/unchanged/dynamic properties.

I kind of expected that we would return the original value, perhaps with a wrapper that added the transient annotations (i.e., DataModificationException).

Perhaps we should provide a helper method (if we don't already) for "wrapping" a TStructuralType instance in an IEdmChangedObject wrapper (that accesses the properties of the existing instance, rather than trying to copy).